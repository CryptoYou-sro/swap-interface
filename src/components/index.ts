export * from './header/header';
export * from './button/button';
export * from './textField/textField';
export * from './swapForm/swapForm';

