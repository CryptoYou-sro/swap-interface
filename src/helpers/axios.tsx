import { useCallback, useEffect, useState } from 'react';
import { Buffer } from 'buffer';
import {
	BASE_URL,
	BINANCE_DEV_URL,
	BINANCE_SCRIPT,
	BIZ_ENTRY_KEY,
	BINANCE_EXCHANGE_INFO,
	BINANCE_PRICE_TICKER
} from './index';
import axios from 'axios';
import destinationNetworks from '../data/destinationNetworks.json';

export enum STATUS_ENUM {
	NONCE = 'NONCE',
	AUTH = 'AUTH',
	PASS = 'PASS'
}

type LocalStorageProps = {
	isKyced: boolean;
	refresh: string;
	access: string;
	account: string;
};

export const apiCall = {
	getNonce: 'nonce?address=',
	auth: 'auth',
	kycToken: 'kyc/token',
	kycStatus: 'kyc/status',
	refresh: 'refresh'
};

export const getMetamaskMessage = (nonce: string): string =>
	`0x${Buffer.from('Please sign this one time nonce: ' + nonce, 'utf8').toString('hex')}`;

export const loadBinanceKycScript = (cb?: any) => {
	const existingId = document.getElementById('binance-kcy-script');

	if (!existingId) {
		const binanceSdkScript = document.createElement('script');
		binanceSdkScript.src = BINANCE_SCRIPT;
		binanceSdkScript.id = 'binance-kcy-script';
		document.body.appendChild(binanceSdkScript);

		binanceSdkScript.onload = () => {
			if (cb) cb();
		};
	}
	if (existingId && cb) cb();
};

export const makeBinanceKycCall = (authToken: string) => {
	// @ts-ignore
	const binanceKyc = new BinanceKyc({
		authToken,
		bizEntityKey: BIZ_ENTRY_KEY,
		apiHost: BINANCE_DEV_URL,
		onMessage: ({ typeCode }: any) => {
			if (typeCode === '102') {
				binanceKyc.switchVisible(true);
			}
		}
	});
};

const getStorageValue = (key: string) => {
	const defaultValue = {
		access: '',
		refresh: '',
		account: '',
		isKyced: false
	} as LocalStorageProps;

	if (typeof window !== 'undefined') {
		const saved = localStorage.getItem(key);

		return saved ? (JSON.parse(saved) as LocalStorageProps) : defaultValue;
	}
};

export const useLocalStorage = (key: string) => {
	const [value, setValue] = useState(() => {
		return getStorageValue(key);
	});

	useEffect(() => {
		localStorage.setItem(key, JSON.stringify(value));
	}, [key, value]);

	return [value, setValue];
};

export const getAuthTokensFromNonce = async (account: string, library: any) => {
	try {
		const res = await axios.request({
			url: `${BASE_URL}${apiCall.getNonce}${account}`
		});
		try {
			const msg = getMetamaskMessage(res.data.nonce);
			const signature = await library?.send('personal_sign', [account, msg]);
			try {
				const tokenRes = await axios.request({
					url: `${BASE_URL}${apiCall.auth}`,
					method: 'POST',
					data: { address: account, signature }
				});

				return tokenRes.data as { access: string; is_kyced: boolean; refresh: string }; // TODO: define Type
			} catch (err: any) {
				throw new Error(err);
			}
		} catch (err: any) {
			throw new Error(err);
		}
	} catch (err: any) {
		throw new Error(err);
	}
};

export const useKyc = (
	authToken: string
): { loading: boolean; error: any; kycStatus: string; kycToken: string } => {
	const [loading, setLoading] = useState(false);
	const [error, setError] = useState(null);
	const [kycStatus, setKycStatus] = useState('');
	const [kycToken, setKycToken] = useState('');

	const fetchData = useCallback(
		async (authToken: string): Promise<void> => {
			try {
				setLoading(true);
				const statusRes = await axios.request({
					// TODO: check typing and if kycStatus is from right place
					url: `${BASE_URL}${apiCall.kycStatus}`,
					headers: {
						Authorization: `Bearer ${authToken}`
					}
				});
				setKycStatus(statusRes.data.statusInfo.kycStatus);
			} catch (err: any) {
				setError(err);
			} finally {
				setLoading(false);
			}

			try {
				setLoading(true);
				const tokenRes = await axios.request({
					url: `${BASE_URL}${apiCall.kycToken}`,
					headers: {
						Authorization: `Bearer ${authToken}`
					}
				});
				setKycToken(tokenRes.data.token);
			} catch (err: any) {
				setError(err);
			} finally {
				setLoading(false);
			}
		},
		[authToken]
	);

	useEffect(() => {
		void fetchData(authToken);
	}, [fetchData]);

	return { loading, error, kycStatus, kycToken };
};

export const useBinanceApi = () => {
	const [allPrices, setAllPrices] = useState<{ symbol: string; price: string }[]>([]);
	const [allFilteredPrices, setAllFilteredPrices] = useState<{ symbol: string; price: string }[]>(
		[]
	);
	const [allPairs, setAllPairs] = useState([]);
	const [allFilteredPairs, setAllFilteredPairs] = useState([]);

	const getExchangeInfo = async () => {
		try {
			const res = await axios.request({ url: BINANCE_EXCHANGE_INFO });
			setAllPairs(res.data.symbols);
		} catch (e: any) {
			throw new Error(e);
		}
	};

	const getTickerData = async () => {
		try {
			const res = await axios.request({ url: BINANCE_PRICE_TICKER });
			setAllPrices(res.data);
		} catch (e: any) {
			throw new Error(e);
		}
	};

	const uniqueTokens: string[] = Object.keys(destinationNetworks).reduce(
		(tokens: string[], network: string) => {
			// @ts-ignore
			const networkTokens = Object.keys(destinationNetworks?.[network]?.['tokens']);

			const allTokens = [...tokens, networkTokens];

			return [...new Set(allTokens.flat(1))];
		},
		['GLMR']
	);

	const isSymbol = (symbol: string): boolean => {
		let k = 0;
		for (let i = 0; i < uniqueTokens.length - 1; i++) {
			k++;
			for (let j = k; j < uniqueTokens.length; j++) {
				if (
					uniqueTokens[i] + uniqueTokens[j] === symbol ||
					uniqueTokens[j] + uniqueTokens[i] === symbol
				) {
					return true;
				}
			}
		}

		return false;
	};

	useEffect(() => {
		void getExchangeInfo();
		void getTickerData();
	}, []);

	useEffect(() => {
		if (allPairs) {
			const filteredPairs = allPairs.filter((pair: any) => isSymbol(pair.symbol));
			setAllFilteredPairs(filteredPairs);
		}

		if (allPrices) {
			const filteredPrices = allPrices.filter((price: any) => isSymbol(price.symbol));
			setAllFilteredPrices(filteredPrices);
		}
	}, [allPairs, allPrices]);

	return { allFilteredPairs, allFilteredPrices };
};
