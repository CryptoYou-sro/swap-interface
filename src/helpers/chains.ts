import { Chain } from '@wagmi/core';
 
export const moonbeam = {
    id: 1284,
    name: 'Moonbeam',
    network: 'moonbeam',
    nativeCurrency: {
        decimals: 18,
        name: 'GLMR',
        symbol: 'GLMR',
    },
    rpcUrls: {
        public: {
            http: ['https://rpc.api.moonbeam.network'],
            webSocket: ['wss://wss.api.moonbeam.network'],
        },
        default: {
            http: ['https://rpc.api.moonbeam.network'],
            webSocket: ['wss://wss.api.moonbeam.network'],
        },
    },
    blockExplorers: {
        default: {
            name: 'Moonscan',
            url: 'https://moonscan.io',
        },
        etherscan: {
            name: 'Moonscan',
            url: 'https://moonscan.io',
        },
    },
    contracts: {
        multicall3: {
            address: '0xcA11bde05977b3631167028862bE2a173976CA11',
            blockCreated: 609002,
        },
    },
    testnet: false
} as const satisfies Chain;

export const bsc = {
    id: 56,
    name: 'Binance Smart Chain',
    network: 'bsc',
    nativeCurrency: {
        decimals: 18,
        name: 'BNB',
        symbol: 'BNB',
    },
    rpcUrls: {
        public: {
            http: ['https://bsc.publicnode.com'],
        },
        default: {
            http: ['https://bsc.publicnode.com'],
        },
    },
    blockExplorers: {
        default: {
            name: 'BscScan',
            url: 'https://bscscan.com',
        },
        etherscan: {
            name: 'BscScan',
            url: 'https://bscscan.com',
        },
    },
    contracts: {
        multicall3: {
            address: '0xca11bde05977b3631167028862be2a173976ca11',
            blockCreated: 15921452,
        },
    },
    testnet: false
} as const satisfies Chain;
