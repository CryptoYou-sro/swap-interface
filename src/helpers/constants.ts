export const LOCAL_STORAGE_THEME = 'darkMode';
export const LOCAL_STORAGE_AUTH = 'auth';

export const INITIAL_STORAGE = {
	access: '',
	refresh: '',
	account: '',
	isKyced: false
};

export const MESSAGE = {
	ignore: 'ignore',
	failed: 'Request failed'
};

export const BIZ_ENTRY_KEY = 'BMYBIT';
export const BINANCE_PROD_URL = 'https://api.commonservice.io';
export const BINANCE_DEV_URL = 'https://dip-qacb.sdtaop.com';
export const BINANCE_SCRIPT =
	'https://static.saasexch.com/static/binance/static/kyc-ui/sdk/0.0.2/sdk.js';
export const BINANCE_PRICE_TICKER = 'https://www.binance.com/api/v3/ticker/price';
export const BINANCE_EXCHANGE_INFO = 'https://api.binance.com/api/v3/exchangeInfo';

export const BASE_URL = 'https://auth-app-aq3rv.ondigitalocean.app/';
// https://auth-app-aq3rv.ondigitalocean.app/account/withdraw/details?id=7ef91f76c79b4793b417e32fbffd73b8
export const MOONBEAM_URL = 'https://rpc.api.moonbeam.network';

export const routes = {
	getNonce: 'nonce?address=',
	auth: 'auth',
	kycToken: 'kyc/token',
	kycStatus: 'kyc/status',
	refresh: 'auth/refresh',
	transactionDetails: 'account/withdraw/details?id='
};

export const PROTOCOL_FEE = 0.002;
export const BINANCE_FEE = 0.002;
export const PROTOCOL_FEE_FACTOR = 1 / (1 - PROTOCOL_FEE);
export const START_TOKEN = 'GLMR';
export const FEE_CURRENCY = 'USDT';

export const ESTIMATED_NETWORK_TRANSACTION_GAS = 55_437;
export const SERVICE_ADDRESS = '0xeB56c1d19855cc0346f437028e6ad09C80128e02';

export const BLOCK_CONTRACT_NUMBER = 2_013_557;
export const WEI_TO_GLMR = 1 / 1000000000000000000;

export const CONTRACT_ADDRESSES = {
	1284: '0x6136F6c8FdC1c711892d217d0D5E95928dce6b20',
	1287: '0x2d249342F1F6D549CE8253782c8A2b9218c68fB2',
	31337: '0x0165878A594ca255338adfa4d48449f69242Eb8F'
};

export const BLOCKS_AMOUNT = 30;

export const makeId = (length: number) => {
	let result = '';
	const characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
	const charactersLength = characters.length;
	for (let i = 0; i < length; i++) {
		result += characters.charAt(Math.floor(Math.random() * charactersLength));
	}

	return result;
};
