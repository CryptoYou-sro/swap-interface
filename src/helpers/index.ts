export * from './functions';
export * from './context';
export * from './constants';
export * from './auth';
export * from './graph';
export * from './storage';
export * from './types';
