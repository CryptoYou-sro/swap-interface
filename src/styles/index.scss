@import "./fonts/font-face.scss";
@import "./variables.scss";
