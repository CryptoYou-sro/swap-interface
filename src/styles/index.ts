export * from './color';
export * from './spacing';
export * from './font';
export * from './mediaQuery';
export * from './types';
