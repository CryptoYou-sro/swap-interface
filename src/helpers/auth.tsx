import { Buffer } from 'buffer';
import { BASE_URL, BINANCE_PROD_URL, BINANCE_SCRIPT, BIZ_ENTRY_KEY } from '../helpers';
import axios from 'axios';
import type { ApiAuthType } from '../helpers';

export enum STATUS_ENUM {
	NONCE = 'NONCE',
	AUTH = 'AUTH',
	PASS = 'PASS'
}

export const routes = {
	getNonce: 'nonce?address=',
	auth: 'auth',
	kycToken: 'kyc/token',
	kycStatus: 'kyc/status',
	refresh: 'auth/refresh'
};

export const getMetamaskMessage = (nonce: string): string =>
	`0x${Buffer.from('Please sign this one time nonce: ' + nonce, 'utf8').toString('hex')}`;

export const loadBinanceKycScript = (cb?: any) => {
	const existingId = document.getElementById('binance-kcy-script');

	if (!existingId) {
		const binanceSdkScript = document.createElement('script');
		binanceSdkScript.src = BINANCE_SCRIPT;
		binanceSdkScript.id = 'binance-kcy-script';
		document.body.appendChild(binanceSdkScript);

		binanceSdkScript.onload = () => {
			if (cb) cb();
		};
	}
	if (existingId && cb) cb();
};

export const makeBinanceKycCall = (authToken: string) => {
	// @ts-ignore
	const binanceKyc = new BinanceKyc({
		authToken,
		bizEntityKey: BIZ_ENTRY_KEY,
		apiHost: BINANCE_PROD_URL,
		onMessage: ({ typeCode }: any) => {
			if (typeCode === '102') {
				binanceKyc.switchVisible(true);
			}
		}
	});
};

export const getAuthTokensFromNonce = async (account: string, library: any) => {
	try {
		const res = await axios.request({
			url: `${BASE_URL}${routes.getNonce}${account}`
		});
		try {
			const msg = getMetamaskMessage(res.data.nonce);
			const signature = await library?.send('personal_sign', [account, msg]);
			try {
				const tokenRes = await axios.request({
					url: `${BASE_URL}${routes.auth}`,
					method: 'POST',
					data: { address: account, signature }
				});

				return tokenRes.data as ApiAuthType;
			} catch (err: any) {
				throw new Error(err);
			}
		} catch (err: any) {
			throw new Error(err);
		}
	} catch (err: any) {
		throw new Error(err);
	}
};
