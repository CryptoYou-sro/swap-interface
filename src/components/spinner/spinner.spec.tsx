import 'jest-styled-components';
import { render } from '@testing-library/react';
import { AuthProvider } from '../../helpers';
import { Spinner } from '../../components';
import { darkTheme } from '../../styles';

describe('Switch', () => {
	it('should render a small spinner with default color', () => {
		const { getByTestId } = render(
			<AuthProvider>
				<Spinner size="small" color={darkTheme.background.default} data-testid="spinner" />
			</AuthProvider>
		);

		expect(getByTestId('spinner')).toMatchSnapshot();
	});

	it('should render a medium spinner with tertiary color', () => {
		const { getByTestId } = render(
			<AuthProvider>
				<Spinner size="medium" color={darkTheme.background.tertiary} data-testid="spinner" />
			</AuthProvider>
		);

		expect(getByTestId('spinner')).toMatchSnapshot();
	});
});
export {};
