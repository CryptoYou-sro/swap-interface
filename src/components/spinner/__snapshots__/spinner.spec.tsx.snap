// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch should match snapshot for a medium sized spinner 1`] = `
.c0 {
  display: inline-block;
  width: 1.5rem;
  height: 1.5rem;
  border: 2px solid;
  border-radius: 50%;
  border-top-color: #161B20;
  -webkit-animation: spin 1s ease-in-out infinite;
  animation: spin 1s ease-in-out infinite;
  -webkit-animation: spin 1s ease-in-out infinite;
}

<div
  class="c0"
  color="#494949"
  data-testid="spinner"
/>
`;

exports[`Switch should match snapshot for a small sized spinner 1`] = `
.c0 {
  display: inline-block;
  width: 1rem;
  height: 1rem;
  border: 2px solid;
  border-radius: 50%;
  border-top-color: #161B20;
  -webkit-animation: spin 1s ease-in-out infinite;
  animation: spin 1s ease-in-out infinite;
  -webkit-animation: spin 1s ease-in-out infinite;
}

<div
  class="c0"
  color="#161B20"
  data-testid="spinner"
/>
`;
